components/ILIAS/OpenIdConnect/classes/class.ilOpenIdConnectSettingsGUI.php

```diff
@@ -672,4 +822,161 @@ protected function setSubTabs(string $active_tab): void
+            $ui_container = $this->buildUserMappingInputFormUDF($definition, $ui_container);
+        }
+        $this->ctrl->setParameter(
+            $this,
+            "opic",
+            "opic_user_data_mapping"
+        );
+        return $this->ui->input()->container()->form()
+                        ->standard(
+                            $this->ctrl->getFormAction($this, 'saveProfile'), $ui_container
+                        );
+    }
+
+    protected function buildUserMappingInputFormUDF($definition, array $ui_container) : array
+    {
+        $value = $this->settings->getProfileMappingFieldValue(self::UDF_STRING . $definition['field_id']);
+        $update = $this->settings->getProfileMappingFieldUpdate(self::UDF_STRING . $definition['field_id']);
+
+        $text_input = $this->ui->input()->field()
+                               ->text($definition['field_name'], '')
+                               ->withValue($value)
+                               ->withDedicatedName(self::UDF_STRING . $definition['field_id'] . self::VALUE_STRING);
+        $checkbox_input = $this->ui->input()->field()->checkbox("", $this->lng->txt('auth_oidc_update_field_info'))
+                                   ->withValue($update)
+                                   ->withDedicatedName(self::UDF_STRING . $definition['field_id'] . self::UPDATE_STRING);
+        $group = $this->ui->input()->field()->group(
+            [$text_input, $checkbox_input]
+        );
+        $ui_container[] = $group;
+
+        return $ui_container;
+    }
+
+    protected function buildUserMappingInputForUserData(string $lang, string $mapping, array $ui_container) : array
+    {
+        $value = $this->settings->getProfileMappingFieldValue($mapping);
+        $update = $this->settings->getProfileMappingFieldUpdate($mapping);
+
+        $text_input = $this->ui->input()->field()
+                               ->text($lang, '')
+                               ->withValue($value)
+                               ->withDedicatedName($mapping . self::VALUE_STRING);
+        $checkbox_input = $this->ui->input()->field()->checkbox("", $this->lng->txt('auth_oidc_update_field_info'))
+                                   ->withValue($update)
+                                   ->withDedicatedName($mapping . self::UPDATE_STRING);
+        $group = $this->ui->input()->field()->group(
+            [$text_input, $checkbox_input]
+        );
+        $ui_container[] = $group;
+
+        return $ui_container;
+    }
+
+    private function initUserDefinedFields(): void
+    {
+        if ($this->udf === null) {
+            $this->udf = ilUserDefinedFields::_getInstance();
+        }
+    }
+
+    /**
+     * @throws ilCtrlException
+     */
+    public function userMapping($form = null): void
+    {
+        if ($form === null) {
+            $form = $this->initUserMappingForm();
+        }
+
+        $request_wrapper = $this->http->wrapper()->query();
+        $active = 2;
```

Comment:
"No magic numbers". Please use constants for this (and for other/similar numbers).